imap-proto/src/parser/rfc4314.rs

```diff
@@ -0,0 +1,109 @@
+//!
+//! Current
+//! https://tools.ietf.org/html/rfc4314
+//!
+//! Original
+//! https://tools.ietf.org/html/rfc2086
+//!
+//! The IMAP ACL Extension
+//!
+
+use nom::{
+    bytes::streaming::tag_no_case,
+    character::complete::{space0, space1},
+    combinator::map,
+    multi::separated_list0,
+    sequence::{preceded, separated_pair, tuple},
+    IResult,
+};
+use std::borrow::Cow;
+
+use crate::parser::core::astring_utf8;
+use crate::parser::rfc3501::mailbox;
+use crate::types::*;
+
+/// 3.6. ACL Response
+/// ```ignore
+/// acl_response  ::= "ACL" SP mailbox SP acl_list
+/// ```
+pub(crate) fn acl(i: &[u8]) -> IResult<&[u8], Response> {
+    let (rest, (_, _, mailbox, acls)) = tuple((
+        tag_no_case("ACL"),
+        space1,
+        map(mailbox, Cow::Borrowed),
+        acl_list,
+    ))(i)?;
+
+    Ok((rest, Response::ACL(ACL { mailbox, acls })))
+}
+
+/// ```ignore
+/// acl_list  ::= *(SP acl_entry)
+/// ```
+fn acl_list(i: &[u8]) -> IResult<&[u8], Vec<ACLEntry>> {
+    preceded(space0, separated_list0(space1, acl_entry))(i)
+}
+
+/// ```ignore
+/// acl_entry ::= SP identifier SP rights
+/// ```
+fn acl_entry(i: &[u8]) -> IResult<&[u8], ACLEntry> {
+    let (rest, (identifier, rights)) = separated_pair(
+        map(astring_utf8, Cow::Borrowed),
+        space1,
+        map(astring_utf8, map_text_to_rights),
+    )(i)?;
+
+    Ok((rest, ACLEntry { identifier, rights }))
+}
+
+/// 3.7. LISTRIGHTS Response
+/// ```ignore
+/// listrights_response  ::= "LISTRIGHTS" SP mailbox SP identifier SP required_rights *(SP optional_rights)
+/// ```
+pub(crate) fn listrights(i: &[u8]) -> IResult<&[u8], Response> {
+    let (rest, (_, _, mailbox, _, identifier, _, required, optional)) = tuple((
+        tag_no_case("LISTRIGHTS"),
+        space1,
+        map(mailbox, Cow::Borrowed),
+        space1,
+        map(astring_utf8, Cow::Borrowed),
+        space1,
+        map(astring_utf8, map_text_to_rights),
+        listrights_optional,
+    ))(i)?;
+
+    Ok((
+        rest,
+        Response::ListRights(ListRights {
+            mailbox,
+            identifier,
+            required,
+            optional,
+        }),
+    ))
+}
+
+fn listrights_optional(i: &[u8]) -> IResult<&[u8], ACLRightList> {
+    let (rest, items) = preceded(space0, separated_list0(space1, astring_utf8))(i)?;
+
+    let v: String = items.into_iter().flat_map(|s| s.chars()).collect();
+
+    Ok((rest, map_text_to_rights(&v)))
+}
+
+pub(crate) fn myrights(i: &[u8]) -> IResult<&[u8], Response> {
+    let (rest, (_, _, mailbox, _, rights)) = tuple((
+        tag_no_case("MYRIGHTS"),
+        space1,
+        map(mailbox, Cow::Borrowed),
+        space1,
+        map(astring_utf8, map_text_to_rights),
+    ))(i)?;
+
+    Ok((rest, Response::MyRights(MyRights { mailbox, rights })))
+}
+
+fn map_text_to_rights(i: &str) -> ACLRightList {
```

Comment:
This should be a method or `From`/`Into` impl on `AclRightList`, instead.

Reply:
I already have the From defined for it. (after extracting it out).  I replaced the uses of this with direct calls to the into() instead.

Reply:
ok.. so with removing of the ACLRightList struct this method is back again as one cannot create a "From" or "Into" implementation between a &str and a Vec<> :(